enhancements/network/persistent-ips-sdn-secondary-networks.md

```diff
@@ -0,0 +1,609 @@
+  delegate IPAM to an external plugin
+- cni conf with the subnet key defined => OVN-K provided IPAM for the provided
+  subnet
+
+**NOTE**: it results in an error if the user attempts to define **both** the
+subnet **and** the IPAM key at the same time.
+
+### Risks and Mitigations
+
+The proposal is assuming the "sticky IPs" feature to be implemented in the IPAM
+plugin - e.g. whereabouts. Performance wise, this might not be a good option
+since we would need to invoke the IPAM plugin via CNI exec based API - i.e.
+invoke the CNI IPAM binary on disk, get the result (the IP), and then use said
+IP in the entire flow. Invoking the IPAM plugin is a synchronous operation that
+would block the OVN-K plugin until it gets the response back.
+
+Given whereabouts instances actively compete for a datastore lock to update the
+IP pool allocations, need to check for the presence of overlapping IPs, and
+would - on top - have to also query the datastore for the presence of existing
+persistent allocations / persist new allocations, it is probable the time it
+takes to provision a pod to grow.
+
+If we implemented the IPAM feature in OVN-Kubernetes instead (that already uses
+an asynchronous model) we would ensure a better throughput when processing the
+CNI requests.
+
+#### Report in the encapsulating object status the IP
+To mitigate the CNI ADD times increase, the controller aware of the
+lifecycle of the encapsulating object (e.g. virt-controller) would persist
+these `PersistentIP`s in the `VirtualMachine.Status`.
+
+Thus, only the initial CNI ADD (VM creation) would take longer: all sub-sequent
+VM start would be faster since the controller would template the pod requesting
+the IP it was allocated.
+
+The following sequence diagrams depict the create / start VM flows - the stop /
+delete flows do not have any changes from their counterparts in
+[Workflow Description](#workflow-description).
+
+##### Create VM
+
+```mermaid
+sequenceDiagram
+
+user ->> KubeVirt: createVM(name=vmA, persistentIPs=true)
+
+KubeVirt ->> Kubernetes: createPod(name=<generated name>, owner=vmA)
+
+note over OVN-K: notices a new pod was added
+OVN-K ->> OVN-K: ip := generateIP()
+
+alt if cniConf.AllowPersistentIPs()
+  OVN-K ->> Kubernetes: persistentIPAllocation(ip, owner=vmA)
+  Kubernetes -->> OVN-K: OK
+end
+Kubernetes -->> KubeVirt: OK
+
+note over KubeVirt: notices pod creation associated for `vmA`
+
+KubeVirt->>KubeVirt: ips := interfaceIPsMap(\n    pod.annotations[NetworkStatusAnnotation]\n)
+
+KubeVirt->>Kubernetes: updateVMStatus(ips...)
+Kubernetes -->>KubeVirt: OK
+
+KubeVirt -->> user: OK
+```
+
+##### Start stopped VM
+```mermaid
+sequenceDiagram
+
+actor user
+
+user ->> KubeVirt: startVM(vm=vmA)
+
+KubeVirt ->> Kubernetes: createPod(<br/>name=<generated name>,<br/>owner=vmA,<br/>ips=vmA.status.ips<br/>)
+
+Kubernetes -->> KubeVirt: OK
+
+note over OVN-K: notices a new pod was added
+
+
+KubeVirt -->> user: OK
```

Comment:
we can achive the same using ovn-k ippool cache and even without that the created persistentIP is already at the informers cache so in reallity the access to api server to read persistentIPs is not going to happend.